File: modules/sql/tables/chunithm.py
from sqlalchemy import (
    Column,
    String,
    Integer,
    BigInteger,
    Float,
    Date,
    CheckConstraint,
    Numeric,
)

from .base import ChunithmMainBase, ChunithmMusicDBBase


class ChunithmBinding(ChunithmMainBase):
    __tablename__ = "bindings"
    im_id = Column(String(50), primary_key=True)
    platform = Column(String(50), primary_key=True)
    server = Column(String(10), primary_key=True)
    aime_id = Column(String(50), nullable=False)


class ChunithmDefaultServer(ChunithmMainBase):
    __tablename__ = "defaults"
    im_id = Column(String(50), primary_key=True)
    platform = Column(String(50), primary_key=True)
    server = Column(String(10), nullable=False)


class ChunithmMusicAlias(ChunithmMainBase):
    __tablename__ = "chunithm_aliases"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    music_id = Column(Integer, nullable=False)
    alias = Column(String(100), nullable=False)


class ChunithmChartData(ChunithmMusicDBBase):
    __tablename__ = "chart_data"
    music_id = Column(Integer, primary_key=True)
    difficulty = Column(Integer, primary_key=True)
    creator = Column(String(50), nullable=True)
    bpm = Column(Float, nullable=True)
    tap_count = Column(Integer, nullable=True)
    hold_count = Column(Integer, nullable=True)
    slide_count = Column(Integer, nullable=True)
    air_count = Column(Integer, nullable=True)
    flick_count = Column(Integer, nullable=True)
    total_count = Column(Integer, nullable=True)


class ChunithmMusic(ChunithmMusicDBBase):
    __tablename__ = "music"
    music_id = Column(Integer, primary_key=True)
    title = Column(String(255), nullable=False)
    artist = Column(String(255), nullable=False)
    category = Column(String(50), nullable=True)
    version = Column(String(10), nullable=True)
    release_date = Column(Date, nullable=True)
    is_deleted = Column(Integer, CheckConstraint("is_deleted IN (0, 1)"), default=0)
    deleted_version = Column(String(10), nullable=True)


class ChunithmMusicDifficulty(ChunithmMusicDBBase):
    __tablename__ = "music_difficulties"
    music_id = Column(Integer, primary_key=True)
    version = Column(String(10), primary_key=True)
    diff0_const = Column(Numeric(3, 1))
    diff1_const = Column(Numeric(3, 1))
    diff2_const = Column(Numeric(3, 1))
    diff3_const = Column(Numeric(3, 1))
    diff4_const = Column(Numeric(3, 1))

File: database/table/chunithm.go
package table

import (
	"time"
)

type ChunithmBinding struct {
	ImID     string `gorm:"column:im_id;type:varchar(30);primaryKey"`
	Platform string `gorm:"column:platform;type:varchar(20);primaryKey"`
	Server   string `gorm:"column:server;type:varchar(10);primaryKey"`
	AimeID   string `gorm:"column:aime_id;type:varchar(50);not null"`
}

func (ChunithmBinding) TableName() string {
	return "bindings"
}

type ChunithmDefaultServer struct {
	ImID     string `gorm:"column:im_id;type:varchar(30);primaryKey"`
	Platform string `gorm:"column:platform;type:varchar(20);primaryKey"`
	Server   string `gorm:"column:server;type:varchar(10);not null"`
}

func (ChunithmDefaultServer) TableName() string {
	return "defaults"
}

type ChunithmMusicAlias struct {
	ID      int64  `gorm:"column:id;primaryKey;autoIncrement"`
	MusicID int    `gorm:"column:music_id;not null"`
	Alias   string `gorm:"column:alias;type:varchar(100);not null"`
}

func (ChunithmMusicAlias) TableName() string {
	return "chunithm_aliases"
}

type ChunithmChartData struct {
	MusicID    int      `gorm:"column:music_id;primaryKey"`
	Difficulty int      `gorm:"column:difficulty;primaryKey"`
	Creator    *string  `gorm:"column:creator;type:varchar(50)"`
	BPM        *float64 `gorm:"column:bpm"`
	TapCount   *int     `gorm:"column:tap_count"`
	HoldCount  *int     `gorm:"column:hold_count"`
	SlideCount *int     `gorm:"column:slide_count"`
	AirCount   *int     `gorm:"column:air_count"`
	FlickCount *int     `gorm:"column:flick_count"`
	TotalCount *int     `gorm:"column:total_count"`
}

func (ChunithmChartData) TableName() string {
	return "chart_data"
}

type ChunithmMusic struct {
	MusicID        int        `gorm:"column:music_id;primaryKey"`
	Title          string     `gorm:"column:title;type:varchar(255);not null"`
	Artist         string     `gorm:"column:artist;type:varchar(255);not null"`
	Category       *string    `gorm:"column:category;type:varchar(50)"`
	Version        *string    `gorm:"column:version;type:varchar(10)"`
	ReleaseDate    *time.Time `gorm:"column:release_date"`
	IsDeleted      int        `gorm:"column:is_deleted;type:int;default:0;check:is_deleted IN (0,1)"`
	DeletedVersion *string    `gorm:"column:deleted_version;type:varchar(10)"`
}

func (ChunithmMusic) TableName() string {
	return "music"
}

type ChunithmMusicDifficulty struct {
	MusicID    int      `gorm:"column:music_id;primaryKey"`
	Version    string   `gorm:"column:version;type:varchar(10);primaryKey"`
	Diff0Const *float64 `gorm:"column:diff0_const;type:numeric(3,1)"`
	Diff1Const *float64 `gorm:"column:diff1_const;type:numeric(3,1)"`
	Diff2Const *float64 `gorm:"column:diff2_const;type:numeric(3,1)"`
	Diff3Const *float64 `gorm:"column:diff3_const;type:numeric(3,1)"`
	Diff4Const *float64 `gorm:"column:diff4_const;type:numeric(3,1)"`
}

func (ChunithmMusicDifficulty) TableName() string {
	return "music_difficulties"
}
